uc-crux-llvm/src/UCCrux/LLVM/Setup.hs

```diff
@@ -0,0 +1,429 @@
+      assumeOne =<< liftIO (LLVMMem.isAligned sym ?ptrWidth regValue alignment)
+    (FTIntRepr w, BVCmp op _ bv) ->
+      assumeOne
+        =<< liftIO
+          ( interpretOp op (LLVMPointer.llvmPointerOffset regValue)
+              =<< W4I.bvLit sym w bv
+          )
+  where
+    assumeOne :: W4I.Pred sym -> Setup m arch sym argTypes (Crucible.RegEntry sym (ToCrucibleType arch atTy))
+    assumeOne pred = assume constraint pred >> pure regEntry
+    interpretOp ::
+      forall w. 1 <= w => L.ICmpOp -> W4I.SymBV sym w -> W4I.SymBV sym w -> IO (W4I.Pred sym)
+    interpretOp =
+      \case
+        L.Ieq -> W4I.bvEq sym
+        L.Ine -> W4I.bvNe sym
+        L.Iult -> W4I.bvUlt sym
+        L.Iule -> W4I.bvUle sym
+        L.Iugt -> W4I.bvUgt sym
+        L.Iuge -> W4I.bvUge sym
+        L.Islt -> W4I.bvSlt sym
+        L.Isle -> W4I.bvSle sym
+        L.Isgt -> W4I.bvSgt sym
+        L.Isge -> W4I.bvSge sym
+
+-- | Generate a fresh symbolic value, recording its \"origin\", i.e. what part
+-- of which global variable or argument it is. That annotation will later be
+-- looked up in "UCCrux.LLVM.Classify.classify" and used to generate additional
+-- preconditions.
+annotatedTerm ::
+  forall m arch sym inTy atTy argTypes.
+  (Crucible.IsSymInterface sym) =>
+  sym ->
+  FullTypeRepr m atTy ->
+  CrucibleTypes.BaseTypeRepr (ToBaseType sym arch atTy) ->
+  -- | Path to this value
+  Selector m argTypes inTy atTy ->
+  Setup m arch sym argTypes (W4I.SymExpr sym (ToBaseType sym arch atTy))
+annotatedTerm sym fullTypeRepr baseTypeRepr selector =
+  do
+    symbol <- freshSymbol
+    (_, expr) <-
+      getAnnotation sym selector fullTypeRepr
+        =<< liftIO (W4I.freshConstant sym symbol baseTypeRepr)
+    pure expr
+
+-- | Like 'annotatedTerm'. Generates a fresh pointer with an annotated block and
+-- offset.
+annotatedLLVMPtr ::
+  forall m arch sym inTy atTy argTypes w.
+  (1 <= w, Crucible.IsSymInterface sym) =>
+  sym ->
+  NatRepr w ->
+  FullTypeRepr m atTy ->
+  -- | Path to this pointer
+  Selector m argTypes inTy atTy ->
+  Setup m arch sym argTypes (LLVMMem.LLVMPtr sym w)
+annotatedLLVMPtr sym w fullTypeRepr selector =
+  do
+    symbol <- freshSymbol
+    ptr <-
+      liftIO . LLVMMem.llvmPointer_bv sym
+        =<< liftIO (W4I.freshConstant sym symbol (W4I.BaseBVRepr w))
+    annotatePointer sym selector fullTypeRepr ptr
+
+pointerRange ::
+  ( ArchOk arch,
+    Crucible.IsSymInterface sym
+  ) =>
+  proxy arch ->
+  sym ->
+  -- | Base pointer
+  LLVMMem.LLVMPtr sym (ArchWidth arch) ->
+  -- | Offset to add
+  W4I.SymBV sym (ArchWidth arch) ->
+  -- | Number of pointers to generate/times to add the offset
+  Int ->
+  IO (NonEmpty (LLVMMem.LLVMPtr sym (ArchWidth arch)))
+pointerRange _proxy sym ptr offset size =
+  reverse
+    <$> foldM
+      ( \(p :| ps) () ->
+          do
+            p' <- LLVMMem.ptrAdd sym ?ptrWidth p offset
+            pure (p' :| (p : ps))
+      )
+      (ptr :| [])
+      (replicate (size - 1) ())
```

Comment:
what if `size == 0`?

Reply:
Great catch, I'll add a panic for now. The goal of adding unfolding operations to parameterized-utils' `Vector` was to make clear the invariant that this returns a list of `size` length, that package just hasn't seen a recent enough release for me to actually use that code here yet.